web3/method.py

```diff
@@ -0,0 +1,144 @@
+    try:
+        method = module_config[attr_name]
+    except KeyError:
+        raise UndefinedMethodError("No method named {0}".format(attr_name))
+    return method_class(module.web3, method)
+
+
+class DummyRequestManager:
+    def request_blocking(method, params):
+        return (method, params)
+
+
+class DummyWeb3:
+    manager = DummyRequestManager
+
+
+def test_sync_method_config_loading():
+    signInBlood = BlockingMethod(DummyWeb3(), TEST_METHOD_CONFIG)
+    signInBlood.input_munger = lambda *_: {}
+    assert signInBlood.method_selector
+    assert ('eth_signTransactionInBlood', {}) == signInBlood()
+
+
+class BaseMethod:
+    """BaseMethod for web3 module methods
+
+    Calls to the Method go through these steps:
+
+    1. input munging ;) - includes normalization, parameter checking, formatters.
+    Any processing on the input parameters that need to happen before json_rpc
+    method string selection occurs.
+
+    2. method selection - function that selects the correct rpc_method. accepts a
+    function or an string.
+
+    3. constructing formatter middlewares - takes the rpc_method and looks up the
+    corresponding input/output formatters. these are the middlewares migrated here.
+
+    4. making the request through the middleware (pipeline)? wrapped request
+    function.
+    """
+    def __init__(self, web3, method_config):
+        self.__name__ = method_config.get('name', 'anonymous')
+        self.__doc__ = method_config.get('doc', '')
+        self.is_property = method_config.get('is_property', False)
+        self.web3 = web3
+        self.input_munger = self._construct_input_pipe(
+            method_config.get('mungers')) or identity
+        self.method_selector = self._method_selector(
+            method_config.get('json_rpc_method'))
+        #  TODO: Write formatter lookup.
+        self.lookup_formatter = None
+
+    def _construct_input_pipe(self, formatters):
+        formatters = formatters or [identity]
+
+    def _method_selector(self, selector):
+        """Method selector can just be the method string.
+        """
+        if isinstance(selector, (str,)):
+            return lambda _: selector
+        else:
+            return selector
+
+    def get_formatters(self, method_string):
+        """Lookup the request formatters for the rpc_method"""
+        if not self.lookup_formatter:
+            return ([identity], [identity],)
+        else:
+            raise NotImplementedError()
+
+    def prep_for_call(self, *args, **kwargs):
+        # takes in input params, steps 1-3
+        params, method, (req_formatters, ret_formatters) = pipe_appends(
+            [self.input_munger, self.method_selector, self.get_formatters],
+            (args, kwargs,))
+        return pipe((method, params,), *req_formatters), ret_formatters
+
+    def __call__(self):
+        raise NotImplementedError()
+
+
+@to_tuple
+def pipe_appends(fns, val):
```

Comment:
Interesting, it's kind of like https://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.accumulate combined with `pipe`.  Maybe re-use the name and call this `accumulate_pipe` or `pipe_and_accumulate`?